### ISSUE TITLE
PathMoverChange Cleanup

### ISSUE THREAD

**Issue by jhprinz** [choderalab/openpathsampling#246]
I spend the last train ride thinking a lot about a more streamlined version of the PathMovers and PathMoveChanges (PMC). This includes also a faster loading which is almost finished.

#### What we have right now
- every PMC takes a sampleset as input
- every PMC has an `ensembles` property to specify on which ensembles the PathMover acts (e.g. RepEx: switch between these two, Forward, shoot in this ensemble, etc...)
- a PMC has a function `apply_to` which will apply the change to a `SampleSet` and return the new changed `SampleSet`
- each PMC has a property `.trials` which returns all used samples in this change in order, this includes rejected samples
- each PMC has a property `.samples` (or renamed `.results`) that contains a list of all accepted samples by the move and are the only ones used in `apply_to`
- the standard `apply_to` method just applies the list of samples from `.samples` to the `SampleSet` to create the new `SampleSet`
- The idea of the PathMovers is to build more complex movers out of simple blocks.
- The simple building blocks include : ShootingMovers, SequentialMovers, RandomChoice, etc...

#### Separate `Sample` generation
The creation of new samples should be put in a separate class. This would still be a mover but a special one. When used as a mover it will pick samples according to the ensembles attribute and do the move. It can also be called as a standalone mover that will generate a new sample from old ones. 
Since this is the only class that will create new samples it will have the connection to the engine and not PathMovers in general. We would call these generators
```
gen = RepExGenerator(ensembles=[ens1, ens2])
pmc = gen.move(globalstate) # will be a mover as usual
samples = gen(samp1, samp2) # will try to swap and return the trial samples, not used for movers
```

Important is that these Generator(-PathMovers) will not accept in the __call__ but use a separate Acceptor (or whatever) to decide if all generated samples will be accepted. So the `__call__` will ONLY generate trials. The Mover accepts or rejects. This means that a `Sample` does not need an `.accepted` property as we are using now. Only the PMC needs to know that. (I think we already agreed on that point a while ago.)

The main goal is to make all generating PathMovers similar and the main `.move` function would look like
```
def move(self, globalstate):
    samples = [globalstate(ens) for ens in self.ensembles] # pick samples
    trials = self(*samples) # call self with these samples
    accepted, details = self.acceptor(trial_samples) # check if accepted and get details
    if accepted:
        return AcceptedPathMoveChange(
            samples=trials,
            mover=self,
            details=details
        )
    else:
        return RejectedPathMoveChange(
            samples=trials,
            mover=self,
            details=details
        )
```

At some point I thought that an acceptor could also be a path mover that takes an underlying mover to generate the trials, but that seemed kind of too much flexibility. The idea behind this is that we could do some generators in a row (without the acceptance) and only accept at the end. Effectively using Movers also for trial generation (like in the minus mover). 

To get this to work we need to do something "dangerous": We need to have PathMovers that will return trials without an acceptance step. This requires that certain mover should only be used inside an acceptance mover and not outside since they do not accept/reject correctly.

I would not add this now (this definitely needs more brain :) 

Generators would be:
- Shooters (Forward, Backward, Two-Way) (ens)
- RepEx (ens1, ens2)
- SubTraj (RandomSubtraj, FirstSubtraj) (initial_ens, sub_ens)
- Extend (Forward, Backward, Two-Way) (initial_ens, extended_ens)

#### Summary
Basically we would keep everything except the .accepted property and streamline the way the moves that generate Samples are coded using `Generators`. That's it.
I would also add an 
- Extend(Forward, Backward) Mover that can then be used in the minus mover, and a
- RestrictToEnsemble Mover that will only return changes that affect the specified ensembles, also for the minus mover.

**Comment by jhprinz** [comment]
### Summary
Since the concept of PathMovers and PathMoveChanges seems so complicated I tried to simplify it:
- In terms of a Replica simulation a SampleSet is the SetOfReplicas that are changed by so-called `PathMoveChanges`

    ```
    SampleSet(1) + Change(1) => SampleSet(2) + Change(2) => SampleSet(3) + ...
    ```

- A Change is simply a list of `Sample`. These samples are used to replace existing samples in the old sampleset in order of appearance. The key to figure out which sample should be replace is the ensemble (although we allow also the replica ID to be used in rare cases. I am not sure if we can avoid using replicas)

    ```
    {samp1(ens1), samp2(ens2)} + [samp3(ens1)] => {samp3(ens1), samp2(ens2)}
    <----- SampleSet(1) -----> + < Change(1) > => <----- SampleSet(2) ----->
    ```

    Important is that it can happen that some samples in a change will be replaced by other samples in the same change and are hence redundant. This is no problem and useful to keep track of intermediate samples. To save storage these intermediate samples could be removed
- All we need to do now is to generate a change and then accept or reject it. This is the simple generalization of single new samples to applying several samples to a set of replicas.

    ```
    [samp3(ens1)]
    < Change(1) > 
    # accepted
    {samp1(ens1), samp2(ens2)} + [samp3(ens1)] => {samp3(ens1), samp2(ens2)}
    <----- SampleSet(1) -----> + < Change(1) > => <----- SampleSet(2) ----->
    # rejected
    {samp1(ens1), samp2(ens2)} + [ NoSamples ] => {samp1(ens1), samp2(ens2)}
    <----- SampleSet(1) -----> + < Change(1) > => <----- SampleSet(1) ----->
    ```

   So, a change will _only return the changing samples if it is accepted_ or _is called accepted if it returns changing samples_ these two things are two sides of the same idea.  

- The instance generating a Change is the PathMover and there are some basic types: 
    1. SampleGeneratingMovers like Shooting, RepEx
    2. DecisionMovers like RandomChoice (pick one mover among many), Conditional (do a mover if another passed)
    3. FlowMovers like Sequential (run several movers sequentially)
- Changes form a semi-group in the sense that there exists always a change that does the same as any sequence of other changes. 
    $ \forall A,B \exists C, s.t. ChangeC = ChangeA + ChangeB$
    So that sequential application of changes is always a valid change
- The `(PathMove)Change` is basically a tree where each node is a list of samples. Since we are not really interested in specific parts of the samples, but mostly in the complete list of changes, we treat the `Change` mainly as a tree and not a list if it comes to python specific functions like `__iter__`, etc...
- The `(PathMove)Change` has some related attributes/properties:

    1. the list of changing samples is contained in `.change` The can also depend on sub-changes
    2. and the list of trial samples (the ones that are applied if accepted are in `.trial`. This can also depend on sub-changes.
    3. any specifically new or generated samples are stored in `.samples` which is a _local_ property and does NOT depend upon sub-changes

### Details (still a little unordered)

- each PM creates a trial change (which are some samples applied in order - as usual) This can be considered as the delta from initial to final step. 
- each PM decides if the change is accepted and thus has an `.accepted` property. This is in accordance with the previous agreement that a move is accepted when it returns samples ( = accepts the trial). 
- for a Sequential PM this means that is always accepts unless in case the trial change is no change (e.g. a sequential move where all submoves were rejected)
- a PM will either accept the complete change or nothing, we do not split changes. This can only be done using submovers
- Sample generation is in general independent of moves and only meant to provide ways to create trial samples. These will then be patched into a trial change.
- Everything that is relevant for the sample creation is stored with each sample, so each sample knows it's origin. This allows tracking like PathTrees and computation of decorrelation times, etc.
- The acceptance (checking if one or multiple samples are valid, rejection because of unsymmetric proposal, etc.) is done by the mover and depends on an acceptor (usually Metropolis)
- Some movers use other changes (list of samples) to generate their change. In this case the more complex move uses the resulting change (which can be empty) and not the trial change
- A PMC has a `.change` property which is an ordered list of `Sample`. These are the samples that will be applied to the `SampleSet` by this change.
- A PMC has a `.trial` property which is an ordered list of `Sample`. These are the samples that are attempted to be applied to the `SampleSet` by this change
- If a PMC was accepted `.change = .trial` otherwise `.change = []`
- In case that we only have a single ensemble all of the above resorts to simple MC, propose a trial sample and accept or reject.
- Every PMC stores its accepted/rejected property although for some complex moves it is purely based upon underlying changes and no random number is rolled. Hence, storage would be redundant. (We might store it but recompute it when it is loaded to make sure!)
- A PM always takes a finite and fixed number of ensembles as input and returns a change that can affect samples in a finite and fixed number of ensembles. 
    1. A Forward Mover takes one particular ensemble as input and outputs (potentially) a change (sample) in the same ensemble A -> A
    2. A RepEx Mover takes two distinct ensembles as input and output 2 samples (or not) in the same ensembles A,B -> A,B
    3. An EnsembleHop will take one ensemble and will return a potential sample in a second different ensemble A -> B
    4. This includes deletion of a sample. The only thing we need is the ensemble
    5. We can have several times the same ensemble, if we require so.
- The specification of input and output ensembles allows effectively parallelization, since we know exactly which moves can depend on which moves: Consider a sequence of ForwardShootings for each ensemble, then from the in/out we can deduce that the next move in the sequence can be run in parallel with the preceding one.
- Finally each PMC `(PathMove)Change` has the following attributes (member variables) and properties (computed attributes)

    ```
    Attributes
    ----------
    mover : PathMover()
        the used PathMover to create this change
    accepted : bool
        `True` if the move is accepted
    subchanges : list of PathMoveChanges()
        a list of used underlying changes, not used in the sample generating moves 
    samples : list of Sample()
        a list of samples created for this move, only used for sample generating moves
    details : Details()
        a details object containing additional information about the change (e.g. for RandomChoice which mover was chosen)
    Properties
    ----------
    accepted : bool
        We will store accepted but potentially only use for some movers, since the more complex once deduce their acceptance from underlying mover results.
    ```

- A PMC will use a special store that allows faster loading and saving.
- All PMs have the following common attributes (member variables) and properties (computed attributes)

    ```
    Attributes
    ----------
    ensembles : dict of Ensemble()
        a move specific description of which ensembles are used for input and output
    submovers : list of PathMover()
        a list of underlying movers used in this move
    Properties
    ----------
    in_ensembles : list of Ensemble()
        a list of Ensemble() that are used for input
    out_ensembles : list of Ensemble()
        a list of Ensemble() that are potentially used for output    
    ```

- A Sample has the following attributes and properties

    ```
    Attributes
    ----------
    trajectory : Trajectory()
        reference to the trajectory of this sample
    ensemble : Ensemble()
        the ensemble this sample should be in or is generated for
    replica : int
        the replica ID that is used for analysis purposes
    parent : Sample()
        the initial sample used to generate this sample
    bias : float
        the proposal bias used in the generation and necessary for acceptance decisions
    mover : PathMover()
        the mover that was used to generate this sample, this has to be a sample generating mover
    details : Details()
        a details object containing additional information specific to the generation of this particular sample (e.g. a shooting contains information about the shooting point, etc...)
    step : int
        a step number (not used anymore?!?) that references the timepjoint when the sample was generated. This information should be contained in the root-mover that is used to generate this particular sample. Since the same sample can appear at various places this can only be the time when the sample was first generated like time of creation.
    Properties
    ----------
    valid : bool
        `True` if the trajectory is in the ensemble
    acceptance : float
        A non-negative number that is zero if the sample is not valid and otherwise the proposal bias. This is usually used to decide if a change is accepted
```

**Comment by dwhswenson** [comment]
More comments coming, but keep in mind that you'll have to change **everything** in analysis as well. And there aren't really tests for analysis, which means merging this will be a headache. If we're not really careful, this will break a lot of stuff.

**Comment by jhprinz** [comment]
You are right, so I tried so far to keep it as contained to the existing API as much as possible. So far the existing tests had been very easy to fix, just some minor adjustments to my surprise and it worked. 

It will also remove lots of old code that is not used anymore

It might be good, if you can merge in as much as possible so that I can take care of the adaption.

In short the only big thing you need to be aware of is, that Samples do not have an accepted property anymore, only the PMCs. That is the only change in the API. I have a few renames, but that is easy to fix.

One thing though: A PathMover has a selection_probability_ratio. Do you use this in analysis or is this legacy? The samples and PMCs have a similar property which I think is more what we want, since the probability depends on the context a sample was generated.

**Comment by dwhswenson** [comment]
Picking out just a couple of things that jumped out at me:

> The key to figure out which sample should be replace is the ensemble (although we allow also the replica ID to be used in rare cases. I am not sure if we can avoid using replicas)

This is incorrect. We should ALWAYS replace based on replica ID, not ensemble. AMS: many replicas in one ensemble. Single replica: one replica, changing ensembles. Replica ID is always a unique key, and the list of active replica IDs is constant over the course of a simulation (until you do some sort of replica spawning/killing approach -- hypothetically interesting, but not part of common path sampling methods). The list of active ensembles is not constant for single replica.

> we treat the Change mainly as a tree and not a list if it comes to python specific functions like `__iter__`, etc

I regularly use `for mover in pmc` and `if mover in pmc` in analysis: so `__iter__` and `__contains__` are pretty important. Both imply that the `Change` is a list of `PathMover`s, which was always a bit strange. If you could give me a list so that this was `mover in pmc.movers` or even `mover in step.movers` (see #244) then I'd be happy.

(More later; posting these first thoughts then replying to your most recent comment above)

**Comment by jhprinz** [comment]
Also: It turns out that the code gets shorter and simpler after the cleanup. I have also added more visualization that demonstrates this:
![retis_scheme](https://cloud.githubusercontent.com/assets/5091934/7609210/fe953842-f970-11e4-8d36-d1c65d229579.png)

**Comment by jhprinz** [comment]
Comment to using replicaID.

True, but we pick the sample that we replace by specifying ensembles (almost always), that is what I actually meant. It is a little confusing to build and explain everything because there might a special case which we want to cover. 

Just to clarify. I keep the old selection / replacing. No changes in SampleSet. Of course all of this requires for our usual cases replicaID to refer to exactly one ensemble. I think for the current status both is fine. But "under the hood" we use replicaID to make sure we are prepared for the future. Since all tests pass while still using the old replace I assume that this works fine.

I guess the main problem is that I always think in ensembles and not ensemble/replica pairs which are most of the time redundant.

Have a look at the picture above, this shows which mover acts on which ensemble. At the bottom you can see that the subtrajectory pickers take a sample from the minus and produce a segment, while the Extender takes a segment and produces a minusensemble. No need for any ForcesEnsemble Changes.
It also turnes out that the Minus only works with innermost interface, segment and minus, etc...

I think this is helpful to see, if everything is working fine. I just realize that my Filter does not display correctly. It should exclude the segment ensemble... Anyway... This is close to being finished from the major recoding of PMC and PM.

**Comment by dwhswenson** [comment]
> In short the only big thing you need to be aware of is, that Samples do not have an accepted property anymore, only the PMCs

I agree that this is the logical way to do it. IIRC, we previously had both. I think I was already using `pmc.accepted`. As I've said before, I look at `Sample`s if I'm calculating ensemble properties, and I look at `Change`s if I'm calculating move-specific properties. The frequency a given sample appears if an ensemble property, but not whether the sample is accepted.

>  A PathMover has a `selection_probability_ratio`. Do you use this in analysis or is this legacy? The samples and PMCs have a similar property which I think is more what we want, since the probability depends on the context a sample was generated.

We need to be very careful in our usage of terminology here. A lot of the early setup used imprecise terms around Monte Carlo, and mixed specific properties of shooting movers with general properties of path movers.

I think `selection_probability_ratio` is a function that is supposed to obtain the ratio of probabilities of selecting a shooting point. It isn't used in any analysis yet, though we'll need information like that (but it is specific to shooting moves). We certainly need to calculate it during the shooting move.

Here's how I think that aspect of things should work (basically, it boils down to Frenkel & Smit):

Properties of every trial we generate, regardless of `PathMover` (personally, I'd put these in `Change.details`):

* `P_gen_old2new`
* `P_gen_new2old`
* `is_acceptable`

The `P_gen_*` are the total probabilities of generating each move (a.k.a. "the underlying matrix of the Markov chain", or simply $\alpha$ in F&S). `is_acceptable` is a boolean (or `1.0`/`0.0`) saying whether the move can be accepted, and is used mainly as a convenience (as I mention below, it can always be seen as an arbitrary bias). For Metropolis MC, the move is then accepted based on

`min(1, is_acceptable * P_gen_old2new / P_get_new2old)`

The `P_gen_*` are made up of individual probabilities, which vary based on the `PathMover` in question. In the most general case, I would argue that the shooting move consists of the following:

* `P_sel` : relative shooting point selection probability
* `P_mod` : relative probability of this modification
* `P_path` : relative probability of generating this specific path
* `P_bias` : arbitrary bias

In practice we tend to make most of those changes isotropic, so under most circumstances only `P_sel` contributes something other than unity to the total ratio. However, as I've mentioned before, I think that there might be some interesting stuff to be done by putting a bias on `P_mod`. (Note also that, technically, `is_acceptable` can be seen as the result of an arbitrary bias.) Each of these has an `_old2new` and a `_new2old` version.

If we're doing a rewrite of `PathMover`s, I think it would be a good time to make the separation between trial generation and acceptance a little clearer. This will allow us to use non-Metropolis acceptance criteria (for example, the waste-recycling approaches which got some attention a few years back).

(Regarding you later comments: thanks for the clarification on `EnsembleFilter` in the minus move; I was confused by that at first.)

**Comment by dwhswenson** [comment]
(tagged with ref docs because this is clearly one of those conversations that we'll want to refer back to when writing documentation!)

**Comment by jhprinz** [comment]
Thanks, that is a very good layout for the acceptance. Right now I have this almost:
`is_acceptable` is just the logical AND of the `.valid` of the samples. gen_old2new and gen_new2old is not available separately since in the Metropolic acceptance only the relative probability is used.

To make things short: I do what you suggest in a simplified way. I generate a sample, e.g. using shooting and store with the sample its gen_old2new/gen_new2old probability (which I call bias, better would be unsymmetric_proposal_bias) this is_acceptable is computed from the sample by checking if it can be in the ensemble it is made for.

Storing the other gen_* attributes is trivial and can be added in a few lines, since the code is now much clearer for these type of changes.

> If we're doing a rewrite of PathMovers, I think it would be a good time to make the separation between trial generation and acceptance a little clearer. This will allow us to use non-Metropolis acceptance criteria (for example, the waste-recycling approaches which got some attention a few years back).

Exactly, this is what I in principle have already, see code of the general 'sample generating' path mover:
```
    def move(self, globalstate):
        # 1. pick a set of ensembles (in case we allow to pick several ones)
        ensembles = self._ensemble_selector()

        # 2. pick samples from these ensembles
        samples = [ self.select_sample(globalstate, ens) for ens in ensembles ]

        # 3. pass these samples to the generator
        trials = self(*samples)

        # 4. accept/reject
        accepted, details = self._accept(trials)

        # 5. and return a PMC
        if accepted:
            return paths.AcceptedSamplePathMoveChange(
                samples=trials,
                mover = self,
                details = details
            )
        else:
            return paths.RejectedSamplePathMoveChange(
                samples=trials,
                mover = self,
                details = details
            )
```

1. We find the ensembles the mover should work with (in case we allow more than one!)
2. We select samples from these ensembles. We can also mix 1 und 2 to make sure that we chose only from non-empty ensembles (but so far we are using only ensemble specific movers)
3. Generate trial samples using the mover specific sample generating method (can also be called directly if needed for whatever reason). These `__call__` function works only in replica ID level and keeps these IDs
4. Push the samples into our acceptor (here check if ALL samples are valid and then multiply the acceptance_probabilities and do an Metropolis)
5. Create a PMC depending on whether it was accepted or not.

If we want to change the acceptance, we put in another function at step 4. So far I hide this and do not store it. But in the future we could have an acceptor class that can be exchanged at will.

**Comment by jhprinz** [comment]
> I regularly use for mover in pmc and if mover in pmc in analysis: so __iter__ and __contains__ are pretty important. Both imply that the Change is a list of PathMovers, which was always a bit strange. If you could give me a list so that this was mover in pmc.movers or even mover in step.movers (see #244) then I'd be happy.

Yes, this is what I meant by saying that it is mainly a tree-like structure and thus iterating will traverse the tree and not run over the actual samples at the given node.

Also checking works now as 
```
pmc # a pathmovechange
fm = ... # a specific e.g. ForwardMover instance
if fm in pmc:
    # if this particular instance is found anywhere in the tree

if ForwardPathMover  in pmc:
    # if any Mover is a subclass of ForwardPathMover

if ShootPathMover in pmc:
    # will be True if a Forward or Backward Mover is in pmc
    # we might change this to `is` and not `subclass` to be more specific
```

**Comment by jhprinz** [comment]
Will be resolved by #251.

**Comment by dwhswenson** [comment]
Question regarding the visualization of movers as depicted above in https://github.com/choderalab/openpathsampling/issues/246#issuecomment-101629542:

You're using letters to label the interfaces. What happens when you have more than 26 interfaces? (For multiple state systems, that's likely to happen.) I assume there are other options for those labels (or options not to label at all) ... one suggestion (stolen from the harp) is to color every Nth vertical line. That keeps those long lines from all looking the same.

**Comment by jhprinz** [comment]
Never thought about that. Right now it uses `char(65 + nr)`, so it will use lots of funny characters, that lower case,etc ... There is surely a better way. Also making the boxes smaller might be good. The idea with colored lines is very good. I will add that.

**Comment by dwhswenson** [comment]
Should have been closed by #251.
